File: java-example/src/main/java/com/kitsune/example/ExamplePlugin.java
package com.kitsune.example;

import com.kitsune.vivid.camera.Camera;
import com.kitsune.vivid.camera.properties.CameraProperties;
import com.kitsune.vivid.camera.properties.DefaultProperties;
import com.kitsune.vivid.motion.MotionPath;
import kotlin.Unit;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerToggleSneakEvent;
import org.bukkit.plugin.java.JavaPlugin;

public class ExamplePlugin extends JavaPlugin implements Listener {

    @Override
    public void onEnable() {

        // register listeners
        Bukkit.getPluginManager().registerEvents(this, this);
    }

    @EventHandler
    public void onPlayerToggleSneakEvent(PlayerToggleSneakEvent event) {

        // only trigger when player is sneaking
        if (event.isSneaking()) return;

        var player = event.getPlayer();
        var camera = new Camera.Builder(this)
                .location(player.getEyeLocation())
                .disconnectHandler(DefaultProperties.DisconnectHandlers.REJOIN)
                .build();
        var start = player.getEyeLocation().clone();

        // set the pitch and yaw
        start.setPitch(90);
        start.setYaw(90);

        // add player to camera
        camera.addViewer(player);

        // create camera motion path
        MotionPath
                .begin()
                .linearPan(start.clone().add(0d, 10d, 0d), 30)
                .forEachViewer(viewer -> {

                    // send this message to each viewer
                    viewer.sendMessage(Component.text("Hello viewer!", NamedTextColor.GRAY));

                    // return Unit.INSTANCE, as we don't have Unit in Java
                    return Unit.INSTANCE;
                })
                .start(camera)// start the motion
                .thenAccept(v -> {
                    camera.destroy(GameMode.CREATIVE);
                })
                .exceptionally(ex -> {

                    // destroy camera if smth goes wrong
                    camera.destroy(GameMode.CREATIVE);
                    return null;
                });

    }
}

File: kotlin-example/src/main/kotlin/com/kitsune/example/ExamplePlugin.kt
package com.kitsune.example

import com.kitsune.vivid.camera.Camera
import com.kitsune.vivid.camera.MotionPath
import com.kitsune.vivid.camera.Interpolation
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.title.TitlePart
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.plugin.java.JavaPlugin

class ExamplePlugin : JavaPlugin(), Listener {

    override fun onEnable() {
        server.pluginManager.registerEvents(this, this)
    }

    @EventHandler
    fun onPlayerJoinEvent(event: PlayerJoinEvent) {

        val camera = Camera(event.player.eyeLocation, this)
        val start = event.player.eyeLocation.clone()

        // add the player as viewer of the camera
        camera.addViewer(event.player)

        MotionPath.begin()
            .linearPan(start.clone().add(10.0, 10.0, 0.0), 80) // linearly pan to new location
            .wait(80) // wait 80 ticks
            .interpolate(start.clone().add(100.0, 10.0, 0.0), 100, Interpolation.SQUARED) // use a different interpolation function
            .switchPosition(start.clone().add(5.0, 0.0, 5.0)) // switch the position of the camera
            .forEachViewer { viewer ->

                // send each viewer of the camera the following title
                viewer.sendTitlePart(TitlePart.TITLE, Component.text("Welcome to the server!", NamedTextColor.GRAY))
            }
            .start(camera) // move the camera along the created path
            .thenAccept { camera.destroy() } // when the animation is complete, destroy the camera
            .exceptionally { exception ->

                // something went wrong!
                logger.warning("Failed to complete camera path!: ${exception.message}")

                null
            }

    }
}